src/npcmove.cpp

```diff
@@ -86,14 +86,73 @@ bool clear_shot_reach( const tripoint &from, const tripoint &to )
     return true;
 }
 
+tripoint good_escape_direction( const npc &who )
+{
+    std::vector<tripoint> candidates;
+
+    const auto rate_pt = [&who]( const tripoint &p ) {
+        if( !g->m.passable( p ) ) {
+            return INT_MAX;
+        }
+
+        int rating = 0;
+        for( const auto &e : g->m.field_at( p ) ) {
+            if( who.is_dangerous_field( e.second ) ) {
+                // @todo Rate fire higher than smoke
+                rating += e.second.getFieldDensity();
+            }
+        }
+
+        return rating;
+    };
+
+    int best_rating = rate_pt( who.pos() );
```

Comment:
Could be refactored as generic STL algorithms:

1. `std::sort` using `rate_pt` as comparator
2. `std::min_element` to find best option
3. `std::remove_if` to eliminate all tiles except the best choices and current tile
4. `random_entry`

Easier to follow and less space for bugs than branching loops